experimental/CliffordAlgOrd/src/CliffordAlgebra.jl

```diff
@@ -0,0 +1,563 @@
+zero(C::CliffordAlgebra) = C()
+
+@doc raw"""
+    one(C::CliffordAlgebra) -> CliffordAlgebraElem
+
+Return the multiplicative identity of the Clifford algebra $C$.
+"""
+function one(C::CliffordAlgebra)
+  res = C()
+  res[1] = base_ring(C)(1)
+  return res
+end
+
+@doc raw"""
+    basis(C::CliffordAlgebra, i::Int) -> CliffordAlgebraElem
+
+Return the $i$-th canonical basis vector of the Clifford algebra $C$.
+"""
+function basis(C::CliffordAlgebra, i::Int)
+  res = C()
+  res[i] = base_ring(C)(1)
+  return res
+end
+
+@doc raw"""
+    gen(C::CliffordAlgebra, i::Int) -> CliffordAlgebraElem
+
+Return the $i$-th canonical algebra generator of the Clifford
+algebra $C$. This is just the image of the $i$-th basis vector
+of the underlying quadratic space under the canonical embedding
+into the Clifford algebra.
+"""
+function gen(C::CliffordAlgebra, i::Int)
+  res = C()
+  if i <= 0
+    throw(BoundsError(res, i))
+  end
+  res[2^(i - 1) + 1] = base_ring(C)(1)
+  return res
+end
+
+@doc raw"""
+    basis(C::CliffordAlgebra) -> Vector{CliffordAlgebraElem}
+
+Return the canonical basis of the Clifford algebra $C$.
+"""
+basis(C::CliffordAlgebra) = map(i -> basis(C, i), 1:dim(C))
+
+@doc raw"""
+    gens(C::CliffordAlgebra) -> Vector{CliffordAlgebraElem}
+
+Return the vector of canonical algebra generators of the Clifford algebra $C$, i.e.,
+if `gram_matrix(C)` is the Gram matrix of the underlying quadratic space with respect
+to the basis (e_1,...,e_n) then the vector of images under the canonical embedding into
+the Clifford algebra is returned.
+"""
+gens(C::CliffordAlgebra) = map(i -> gen(C, i), 1:_dim_qf(C))
+
+@doc raw"""
+    is_commutative(C::CliffordAlgebra) -> Bool
+
+Return `true` if $C$ is commutative and `false` otherwise.
+"""
+is_commutative(C::CliffordAlgebra) = dim(C) == 1 || dim(C) == 2
+
+################################################################################
+#
+#  Element Access
+#
+################################################################################
+
+@doc raw"""
+    coeff(x::CliffordAlgebraElem, i::Int) -> FieldElem
+
+Return the `i`-th coefficient of the element `x`.
+"""
+coeff(x::CliffordAlgebraElem, i::Int64) = coefficients(x)[i]
+
+getindex(x::CliffordAlgebraElem, i::Int64) = coefficients(x)[i]
+
+function setindex!(x::CliffordAlgebraElem, newentry::RingElement, i::Int64) 
+  coefficients(x)[i] = newentry
+end
+
+################################################################################
+#
+#  Other functionality
+#
+################################################################################
+
+@doc raw"""
+    basis_of_centroid(C::CliffordAlgebra) -> Vector{CliffordAlgebraElem}
+
+Return a basis of the centroid of $C$. Unless `dim(space(C)) = 0`, it consists of
+two elements. The first one is the multiplicative identity of $C$. The square of
+the second basis element, if present, equals `quadratic_discriminant(C)`.
+"""
+function basis_of_centroid(C::CliffordAlgebra)::Vector{elem_type(C)}
```

Comment:
```suggestion
function basis_of_centroid(C::CliffordAlgebra)
```
This lowers to a type conversion. To get a type assertion, instead put this to each line with a return, eg here `return C.basis_of_centroid::Vector{elem_type(C)}`. (at all places you just modified

Reply:
I did not know that, thanks